### ISSUE TITLE
RuntimeError: clEnqueueNDRangeKernel failed: OUT_OF_RESOURCES

### ISSUE THREAD

**Issue by StefaniaOliviero** [robbert-harms/MDT#33]
Good morning,
I am using MDT to fit synthetic signals coming from a Monte Carlo diffusion simulation in synthetic brain tissues. I am successfully fitting the signals when using acquisition protocols with 300 or 600 acquisitions, but, unfortunately, I am getting the following error, when using sequences with 2000 or 3000 acquisitions. I don't understand if the problem concerns the requested RAM  or the processor: in both the cases of 300 and 600 acquisitions the RAM used is less than 3 GB, while the available RAM in all the cases is more than 50 GB.

Thanks in advance for any help.

[soliviero@titanic connectoma300_plus]$ mdt-model-fit AxCaliber B_p00_g70.nii connectoma300_plus.prtcl MDT_mask_scaled.nii -n 0.05 --method Levenberg-Marquardt -o outputLM/B_p00_g70 --cl-device-ind 0
/usr/local/lib/python3.6/site-packages/mdt/protocols.py:627: RuntimeWarning: invalid value encountered in true_divide
  (protocol['Delta'] - (protocol['delta'] / 3.0))))
[2021-02-25 10:55:12,601] [INFO] [mdt.lib.processing.model_fitting] [get_model_fit] - Starting intermediate optimization for generating initialization point.
[2021-02-25 10:55:12,755] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Using MDT version 1.2.6
[2021-02-25 10:55:12,755] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Preparing for model BallStick_r1
[2021-02-25 10:55:12,776] [INFO] [mdt.models.composite] [_prepare_input_data] - No volume options to apply, using all 3913 volumes.
[2021-02-25 10:55:12,776] [INFO] [mdt.lib.processing.model_fitting] [_model_fit_logging] - Fitting BallStick_r1 model
[2021-02-25 10:55:12,776] [INFO] [mdt.lib.processing.model_fitting] [_model_fit_logging] - The 4 parameters we will fit are: ['S0.s0', 'w_stick0.w', 'Stick0.theta', 'Stick0.phi']
[2021-02-25 10:55:12,776] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Saving temporary results in outputLM/B_p00_g70/BallStick_r1/tmp_results.
[2021-02-25 10:55:12,936] [INFO] [mdt.lib.processing.processing_strategies] [_process_chunk] - Computations are at 0.00%, processing next 16 voxels (16 voxels in total, 0 processed). Time spent: 0:00:00:00, time left: ? (d:h:m:s).
[2021-02-25 10:55:12,936] [INFO] [mdt.lib.processing.model_fitting] [_process] - Starting optimization
[2021-02-25 10:55:12,936] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using MOT version 0.11.3
[2021-02-25 10:55:12,936] [INFO] [mdt.lib.processing.model_fitting] [_process] - We will use a single precision float type for the calculations.
[2021-02-25 10:55:12,937] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using device 'GPU - TITAN V (NVIDIA CUDA)'.
[2021-02-25 10:55:12,937] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using compile flags: ('-cl-denorms-are-zero', '-cl-mad-enable', '-cl-no-signed-zeros')
[2021-02-25 10:55:12,937] [INFO] [mdt.lib.processing.model_fitting] [_process] - We will use the optimizer Levenberg-Marquardt with default settings.
Traceback (most recent call last):
  File "/usr/local/bin/mdt-model-fit", line 8, in <module>
    sys.exit(ModelFit.console_script())
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/shell_utils.py", line 47, in console_script
    cls().start(sys.argv[1:])
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/shell_utils.py", line 66, in start
    self.run(args, {})
  File "/usr/local/lib/python3.6/site-packages/mdt/cli_scripts/mdt_model_fit.py", line 161, in run
    fit_model()
  File "/usr/local/lib/python3.6/site-packages/mdt/cli_scripts/mdt_model_fit.py", line 155, in fit_model
    use_cascaded_inits=args.use_cascaded_inits)
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 191, in fit_model
    double_precision=double_precision)
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 99, in get_optimization_inits
    double_precision=double_precision)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 177, in get_optimization_inits
    return get_init_data(model_name)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 161, in get_init_data
    fit_results = get_model_fit('BallStick_r1')
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 86, in get_model_fit
    cl_device_ind=cl_device_ind, initialization_data={'inits': inits})
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 206, in fit_model
    optimizer_options=optimizer_options)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 306, in fit_composite_model
    return processing_strategy.process(worker)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 106, in process
    self._process_chunk(processor, chunks)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 151, in _process_chunk
    process()
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 148, in process
    processor.process(chunk, next_indices=next_chunk)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 275, in process
    self._process(roi_indices, next_indices=next_indices)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 380, in _process
    options=self._optimizer_options)
  File "/usr/local/lib/python3.6/site-packages/mot/optimize/__init__.py", line 123, in minimize
    constraints_func=constraints_func, data=data, options=options)
  File "/usr/local/lib/python3.6/site-packages/mot/optimize/__init__.py", line 532, in _minimize_levenberg_marquardt
    cl_runtime_info=cl_runtime_info)
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_function.py", line 362, in evaluate
    events = processor.process(wait_for=wait_for)
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_processors.py", line 90, in process
    events.update(worker.process(wait_for=wait_for))
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_processors.py", line 144, in process
    wait_for=wait_for)
  File "/usr/local/lib64/python3.6/site-packages/pyopencl/__init__.py", line 861, in kernel_call
    return self._enqueue(self, queue, global_size, local_size, *args, **kwargs)
  File "<generated code>", line 189, in enqueue_knl_kernel_lmmin
pyopencl._cl.RuntimeError: clEnqueueNDRangeKernel failed: OUT_OF_RESOURCES
[soliviero@titanic connectoma300_plus]$ mdt-model-fit AxCaliber B_p00_g70.nii connectoma300_plus.prtcl MDT_mask_scaled.nii -n 0.05 --method Levenberg-Marquardt -o outputLM/B_p00_g70 --cl-device-ind 0
/usr/local/lib/python3.6/site-packages/mdt/protocols.py:627: RuntimeWarning: invalid value encountered in true_divide
  (protocol['Delta'] - (protocol['delta'] / 3.0))))
[2021-02-26 12:10:00,522] [INFO] [mdt.lib.processing.model_fitting] [get_model_fit] - Starting intermediate optimization for generating initialization point.
[2021-02-26 12:10:00,678] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Using MDT version 1.2.6
[2021-02-26 12:10:00,696] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Preparing for model BallStick_r1
[2021-02-26 12:10:00,702] [INFO] [mdt.models.composite] [_prepare_input_data] - No volume options to apply, using all 3913 volumes.
[2021-02-26 12:10:00,703] [INFO] [mdt.lib.processing.model_fitting] [_model_fit_logging] - Fitting BallStick_r1 model
[2021-02-26 12:10:00,703] [INFO] [mdt.lib.processing.model_fitting] [_model_fit_logging] - The 4 parameters we will fit are: ['S0.s0', 'w_stick0.w', 'Stick0.theta', 'Stick0.phi']
[2021-02-26 12:10:00,703] [INFO] [mdt.lib.processing.model_fitting] [fit_composite_model] - Saving temporary results in outputLM/B_p00_g70/BallStick_r1/tmp_results.
[2021-02-26 12:10:00,865] [INFO] [mdt.lib.processing.processing_strategies] [_process_chunk] - Computations are at 0.00%, processing next 16 voxels (16 voxels in total, 0 processed). Time spent: 0:00:00:00, time left: ? (d:h:m:s).
[2021-02-26 12:10:00,865] [INFO] [mdt.lib.processing.model_fitting] [_process] - Starting optimization
[2021-02-26 12:10:00,866] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using MOT version 0.11.3
[2021-02-26 12:10:00,866] [INFO] [mdt.lib.processing.model_fitting] [_process] - We will use a single precision float type for the calculations.
[2021-02-26 12:10:00,866] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using device 'GPU - TITAN V (NVIDIA CUDA)'.
[2021-02-26 12:10:00,866] [INFO] [mdt.lib.processing.model_fitting] [_process] - Using compile flags: ('-cl-denorms-are-zero', '-cl-mad-enable', '-cl-no-signed-zeros')
[2021-02-26 12:10:00,866] [INFO] [mdt.lib.processing.model_fitting] [_process] - We will use the optimizer Levenberg-Marquardt with default settings.
Traceback (most recent call last):
  File "/usr/local/bin/mdt-model-fit", line 8, in <module>
    sys.exit(ModelFit.console_script())
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/shell_utils.py", line 47, in console_script
    cls().start(sys.argv[1:])
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/shell_utils.py", line 66, in start
    self.run(args, {})
  File "/usr/local/lib/python3.6/site-packages/mdt/cli_scripts/mdt_model_fit.py", line 161, in run
    fit_model()
  File "/usr/local/lib/python3.6/site-packages/mdt/cli_scripts/mdt_model_fit.py", line 155, in fit_model
    use_cascaded_inits=args.use_cascaded_inits)
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 191, in fit_model
    double_precision=double_precision)
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 99, in get_optimization_inits
    double_precision=double_precision)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 177, in get_optimization_inits
    return get_init_data(model_name)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 161, in get_init_data
    fit_results = get_model_fit('BallStick_r1')
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 86, in get_model_fit
    cl_device_ind=cl_device_ind, initialization_data={'inits': inits})
  File "/usr/local/lib/python3.6/site-packages/mdt/__init__.py", line 206, in fit_model
    optimizer_options=optimizer_options)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 306, in fit_composite_model
    return processing_strategy.process(worker)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 106, in process
    self._process_chunk(processor, chunks)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 151, in _process_chunk
    process()
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 148, in process
    processor.process(chunk, next_indices=next_chunk)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/processing_strategies.py", line 275, in process
    self._process(roi_indices, next_indices=next_indices)
  File "/usr/local/lib/python3.6/site-packages/mdt/lib/processing/model_fitting.py", line 380, in _process
    options=self._optimizer_options)
  File "/usr/local/lib/python3.6/site-packages/mot/optimize/__init__.py", line 123, in minimize
    constraints_func=constraints_func, data=data, options=options)
  File "/usr/local/lib/python3.6/site-packages/mot/optimize/__init__.py", line 532, in _minimize_levenberg_marquardt
    cl_runtime_info=cl_runtime_info)
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_function.py", line 362, in evaluate
    events = processor.process(wait_for=wait_for)
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_processors.py", line 90, in process
    events.update(worker.process(wait_for=wait_for))
  File "/usr/local/lib/python3.6/site-packages/mot/lib/cl_processors.py", line 144, in process
    wait_for=wait_for)
  File "/usr/local/lib64/python3.6/site-packages/pyopencl/__init__.py", line 861, in kernel_call
    return self._enqueue(self, queue, global_size, local_size, *args, **kwargs)
  File "<generated code>", line 189, in enqueue_knl_kernel_lmmin
pyopencl._cl.RuntimeError: clEnqueueNDRangeKernel failed: OUT_OF_RESOURCES
[soliviero@titanic connectoma300_plus]$

**Comment by robbert-harms** [comment]
Hi StefaniaOliviero,

Thanks for reaching out. The problem is the number of acquisitions (2000~3000) in combination with the Levenberg-Marquardt (LM) algorithm. To work, the LM algorithm requires the Jacobian matrix which is a n-times-n matrix with n the number of acquisitions. As you can see, this matrix scales quadratic with increasing acquisitions. For smaller datasets this fits in the memory of your GPU, for larger n it no longer fits. 

Basically the only suggestion I have is to run the computations using a different algorithm, Powell or Nelder mead Simplex.

I am aware of this memory issue and I hope to address it in future versions of MDT. At the time when I started with MDT (~8 years ago), such large datasets were not yet common. We typically dealt with ~60 acquisitions, not ~2000.

Best wishes,

Robbert

**Comment by StefaniaOliviero** [comment]
Hi Robbert
Thanks a lot for answering so quickly. 

Best wishes to you
Stefania Oliviero
